### PR TITLE
[main] add `Application` class

### PR DESCRIPTION
**Pull request by simonlingoogle**
This commit adds the `Application` class for better managing different program components.

### CODE REVIEW

File: src/agent/program.cpp [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    // allow quitting elegantly
+    signal(SIGTERM, HandleSignal);
+
+    while (!sShouldTerminate)
+    {
+        otbr::MainloopContext mainloop;
+        int                   rval;
+
+        mainloop.mMaxFd   = -1;
+        mainloop.mTimeout = kPollTimeout;
```
Comment (bukepo) [inline src/agent/program.cpp:92]: Should we consider remove this default timeout? If there's no no scheduled tasks, select() blocks until an event occurs.

Comment (simonlingoogle) [reply src/agent/program.cpp:93]: I agree. But may leave for another PR.

**Comment by simonlingoogle** [comment]
Should we name it `Program` or `Application` or `OtbrAgent`? @bukepo @wgtdkp @superwhd

**Comment by wgtdkp** [comment]
> This commit adds the `Program` class for better managing different program components.

Why do we need this commit? Are there more "program" incoming?

File: Android.mk [hunk 2]
```diff
@@ -107,6 +107,7 @@ LOCAL_CPPFLAGS += -std=c++14
 LOCAL_GENERATED_SOURCES = $(OTBR_GEN_DBUS_INTROSPECT_HEADER)
 
 LOCAL_SRC_FILES := \
+	src/agent/program.cpp \
```
Comment (superwhd) [inline Android.mk:110]:
```suggestion
    src/agent/program.cpp \
```

File: src/agent/program.hpp [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+ *   This module includes definition for OTBR program.
+ *
+ * @{
+ */
+
+/**
+ * This class implements OTBR program functionality.
+ *
+ */
+class Program
```
Comment (superwhd) [inline src/agent/program.hpp:76]: What about name it as `Application` or `App`?

**Comment by superwhd** [comment]
> Should we name it `Program` or `Application` or `OtbrAgent`? @bukepo @wgtdkp @superwhd

I prefer `Application`.

**Comment by bukepo** [comment]
> > Should we name it `Program` or `Application` or `OtbrAgent`? @bukepo @wgtdkp @superwhd
> 
> I prefer `Application`.

+1

File: src/agent/program.hpp (hunk 3 above)
Comment (simonlingoogle) [reply src/agent/program.hpp]: DONE

File: Android.mk [hunk 4]
```diff
@@ -107,6 +107,7 @@ LOCAL_CPPFLAGS += -std=c++14
 LOCAL_GENERATED_SOURCES = $(OTBR_GEN_DBUS_INTROSPECT_HEADER)
 
 LOCAL_SRC_FILES := \
+	src/agent/application.cpp \
```
Comment (simonlingoogle) [inline Android.mk:110]:
```suggestion
    src/agent/application.cpp \
```

**Review by wgtdkp — APPROVED** [review]
LGTM 👍

File: src/agent/application.hpp [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+#include <signal.h>
+#include <stdint.h>
+#include <vector>
+
+#include "border_agent/border_agent.hpp"
+#include "ncp/ncp_openthread.hpp"
+
+#if OTBR_ENABLE_REST_SERVER
+#include "rest/rest_web_server.hpp"
+using otbr::rest::RestWebServer;
```
Comment (wgtdkp) [inline src/agent/application.hpp:46]: Do not use `using` in hdeader files. May use it inside class `Application`.

Comment (superwhd) [reply src/agent/application.hpp:46]: +1

**Review by superwhd — APPROVED** [review]
LGTM over all

File: src/agent/application.hpp [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+
+/**
+ * This class implements OTBR application management.
+ *
+ */
+class Application
+{
+public:
+    /**
+     * This constructor initialize the Application instance.
```
Comment (bukepo) [inline src/agent/application.hpp:76]:
```suggestion
     * This constructor initializes the Application instance.
```

Comment (bukepo) [reply src/agent/application.hpp:76]: Is this kind of documenting necessary?

Comment (wgtdkp) [reply src/agent/application.hpp:76]: +1

Comment (simonlingoogle) [reply src/agent/application.hpp:76]: may keep it since the parameters are documented too.

**Review by bukepo — APPROVED** [review]

File: src/agent/application.cpp [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+
+/**
+ * @file
+ *   The file implements the OTBR Agent.
+ */
+
+#define OTBR_LOG_TAG "APP"
+
+#include "agent/application.hpp"
+#include <common/mainloop_manager.hpp>
```
Comment (jwhui) [inline src/agent/application.cpp:37]:
```suggestion
#include "common/mainloop_manager.hpp"
```

**Review by jwhui — APPROVED** [review]

**Pull request merged**